src/cachemachine/rubintagfuncs.py

```diff
@@ -0,0 +1,259 @@
+"""Functions called from the classmethod that builds a RubinTag object.
+Because we don't have an object yet, these are functions, rather than methods.
+"""
+
+import re
+from typing import Dict, List, Match, Optional, Tuple
+
+import structlog
+from semver import VersionInfo
+
+from cachemachine.rubintagtypes import RubinTagType
+
+DOCKER_DEFAULT_TAG = "latest"
+logger = structlog.get_logger(__name__)
+
+# Build the raw strings for tags and tag components, which can then be
+# mixed and matched to some degree.  Cuts down a little on the complexity of
+# the TAGTYPE_REGEXPS tuple and prevents some duplication.  We will use named
+# group matches to make understanding the tag data easier.
+TAG: Dict[str, str] = {
+    "release": r"r(?P<major>\d+)_(?P<minor>\d+)_(?P<patch>\d+)",
+    "rc": r"r(?P<major>\d+)_(?P<minor>\d+)_(?P<patch>\d+)_rc(?P<pre>\d+)",
+    "weekly": r"w_(?P<year>\d+)_(?P<week>\d+)",
+    "daily": r"d_(?P<year>\d+)_(?P<month>\d+)_(?P<day>\d+)",
+    "experimental": r"exp_(?P<rest>.*)",
+    "cycle": r"_(?P<ctag>c|csal)(?P<cycle>\d+\.\d+)",
```

Comment:
Why is cycle a floating point number? What does the bit after the period mean? Is it semantically part of the cycle and thus part of what we specify as the current cycle to exclude things from other cycles?

Reply:
The real answer is "because that's the way Tiago is doing it."  It is the cycle number followed by a build number, which in Tiago's build method is a sequentially-increasing integer.

Reply:
done